fletcher/base.py

```diff
@@ -262,47 +256,57 @@ def __setitem__(self, key, value):
-                    arr = np.array(chunk.to_pylist())
-                    arr[array_chunk_indices] = np.array(value)[key_chunk_indices]
-                    chunks.append(pa.array(arr, self.dtype.arrow_dtype))
-                else:
-                    arr = chunk.to_pandas()
-                    # In the case where we zero-copy Arrow to Pandas conversion, the
-                    # the resulting arrays are read-only.
-                    if not arr.flags.writeable:
-                        arr = arr.copy()
-                    arr[array_chunk_indices] = np.array(value)[key_chunk_indices]
-                    # ARROW-2806: Inconsistent handling of np.nan requires adding a mask
-                    if pa.types.is_integer(
-                        self.dtype.arrow_dtype
-                    ) or pa.types.is_floating(self.dtype.arrow_dtype):
-                        mask = pd.isna(arr)
-                    else:
-                        mask = None
-                    chunks.append(pa.array(arr, self.dtype.arrow_dtype, mask=mask))
+        affected_chunks_index = self._get_chunk_indexer(key)
+        affected_chunks_unique = np.unique(affected_chunks_index)
+
+        all_chunks = list(self.data.iterchunks())
+
+        for ix, offset in zip(
+            affected_chunks_unique, self.offsets[affected_chunks_unique]
+        ):
+            chunk = all_chunks[ix]
+
+            # Translate the array-wide indices to indices of the chunk
+            key_chunk_indices = np.argwhere(affected_chunks_index == ix).flatten()
+            array_chunk_indices = key[key_chunk_indices] - offset
+
+            if pa.types.is_date64(self.dtype.arrow_dtype):
+                # ARROW-2741: pa.array from np.datetime[D] and type=pa.date64 produces invalid results
+                arr = np.array(chunk.to_pylist())
+                arr[array_chunk_indices] = np.array(value)[key_chunk_indices]
+                pa_arr = pa.array(arr, self.dtype.arrow_dtype)
             else:
-                chunks.append(chunk)
-            offset += len(chunk)
+                arr = chunk.to_pandas()
+                # In the case where we zero-copy Arrow to Pandas conversion, the
+                # the resulting arrays are read-only.
+                if not arr.flags.writeable:
+                    arr = arr.copy()
+
+                arr[array_chunk_indices] = value[key_chunk_indices]
+
+                # ARROW-2806: Inconsistent handling of np.nan requires adding a mask
+                if pa.types.is_integer(self.dtype.arrow_dtype) or pa.types.is_floating(
+                    self.dtype.arrow_dtype
+                ):
+                    mask = pd.isna(arr)
+                else:
+                    mask = None
+                pa_arr = pa.array(arr, self.dtype.arrow_dtype, mask=mask)
+            all_chunks[ix] = pa_arr
 
-        self.data = pa.chunked_array(chunks)
+        self.data = pa.chunked_array(all_chunks)
```

Comment:
After setting data, you should recalculate the offsets. Maybe make `data` a property and implement a setter that always recalculates them on assignment.

Reply:
setitem doesn't change the size of the chunks. the only operation we currently support which changes the offsets is concat which calls the constructor.